src/entities/causalChain.ts

```diff
@@ -0,0 +1,263 @@
+            });
+            chain += "] => (";
+            chain += link.toString() + ", " + link.getEvent();
+            chain += ")\n";
+        });
+
+        return chain;
+    }
+}
+
+/**
+ * A link in a causal chain, representing an event of type T,
+ * with causes of type S and effects also of type S.
+ *  TODO: At the moment, there is no recycling of identifiers,
+ *  so this number will grow ad infinitum.
+ */
+export class Link<S, T> {
+
+    private static counter: number = 0;
+    private identifier: number;
+    private ingress: Collections.Set<Link<S, T>>;
+    private egress: Collections.Set<Link<S, T>>;
+
+    /**
+     * Create a new Link.
+     * @param event
+     * @param before
+     * @param after
+     */
+    public constructor(private readonly event: T,
+                       private readonly before: Immutable.Set<S>,
+                       private readonly after: Immutable.Set<S>) {
+        this.identifier = Link.counter;
+        Link.counter++;
+        this.ingress = new Collections.Set<Link<S, T>>();
+        this.egress = new Collections.Set<Link<S, T>>();
+    }
+
+    /**
+     * The causes.
+     * @returns {Immutable.Set<S>}
+     */
+    public get causes(): Immutable.Set<S> {
+        return this.before;
+    }
+
+    /**
+     * The effects.
+     * @returns {Immutable.Set<S>}
+     */
+    public get effects(): Immutable.Set<S> {
+        return this.after;
+    }
+
+    /**
+     * The causally antecedent links.
+     * @returns {Set<Link<S, T>>}
+     */
+    public get antecedent(): Collections.Set<Link<S, T>> {
+        return this.ingress;
+    }
+
+    /**
+     * The causally incident links.
+     * @returns {Set<Link<S, T>>}
+     */
+    public get incident(): Collections.Set<Link<S, T>> {
+        return this.egress;
+    }
+
+    /**
+     * Retrieve the event represented by this link.
+     * @returns {T}
+     */
+    public getEvent(): T {
+        return this.event;
+    }
+
+    /**
+     * Link this chain to another chain,
+     * which will then exhibit causal incidence.
+     * Returns whether or not the act of linking was successful.
+     * @param {Link<S, T>} other
+     * @returns {boolean}
+     */
+    public linkTo(other: Link<S, T>): boolean {
+        let reachable = false;
+        this.effects.forEach(c => {
+            if (c !== undefined && other.causes.contains(c)) {
+                reachable = true;
+            }
+        });
+
+        if (reachable) {
+            this.incident.add(other);
+            other.antecedent.add(this);
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Find an ancestor of this link with an event identical to that of a given link.
+     * Returns whether a similar ancestor was found.
+     * @param {Link<S, T>} toMatch
+     * @returns {Link<S, T> | undefined}
+     */
+    public hasSimilarAncestor(toMatch: Link<S, T>): boolean {
```

Comment:
Return type here doesn't match the `@returns {Link<S, T> | undefined}` doc comment. You know, you actually don't have to include types in the docs. Duplicating this information will only make it go out of sync.

Reply:
Yeah, my IDE does that automatically, and the incorrect @returns is indeed due to some hasty refactoring. I'll make sure to change it.